opm/autodiff/BlackoilModelBase_impl.hpp

```diff
@@ -1121,6 +1167,88 @@ namespace detail {
+                                                              SolutionState& state,
+                                                              WellState& well_state)
+    {
+        V aliveWells;
+        const int np = wells().number_of_phases;
+        std::vector<ADB> cq_s(np, ADB::null());
+
+        int          it  = 0;
+        std::vector<V> vars0;
+        //bhp and Q for the wells
+        vars0.reserve(2);
+        variableWellStateInitials(well_state,vars0);
+        std::vector<ADB> vars = ADB::variables(vars0);
+        std::vector<int> indices = variableWellsStateIndices();
+        SolutionState state0 = state;
+        asImpl().makeConstantState(state0);
+        SolutionState wellSolutionState = state0;
+        variableStateExtractWellsVars(indices,vars,wellSolutionState);
+        std::vector<ADB> mob_perfcells_const(np, ADB::null());
+        std::vector<ADB> b_perfcells_const(np, ADB::null());
+        for (int phase = 0; phase < np; ++phase) {
+            mob_perfcells_const[phase] = ADB::constant(mob_perfcells[phase].value());
+            b_perfcells_const[phase] = ADB::constant(b_perfcells[phase].value());
+        }
+        asImpl().addWellEq(wellSolutionState, well_state, mob_perfcells_const, b_perfcells_const, aliveWells,cq_s);
+        addWellControlEq(wellSolutionState, well_state, aliveWells);
+        bool converged = getWellConvergence(it);
+        while ( (!converged && (it < 15))) {
+
+            std::vector<ADB> eqs;
+            eqs.reserve(2);
+            eqs.push_back(residual_.well_flux_eq);
+            eqs.push_back(residual_.well_eq);
+            ADB total_residual = vertcatCollapseJacs(eqs);
+            const std::vector<M>& Jn = total_residual.derivative();
+            const Eigen::SparseLU< M > solver(Jn[0]);
+            const Eigen::VectorXd& dx = solver.solve(total_residual.value().matrix());
+            const int numeq = well_state.numWells()*(well_state.numPhases()+1);
+            V dx_V = V(numeq);
+            std::copy_n(dx.data(),numeq, dx_V.data());
+            updateWellState(dx_V,well_state);
+            updateWellControls(well_state);
+            //bhp and Q for the wells
+            vars0.clear();
+            variableWellStateInitials(well_state,vars0);
+            vars = ADB::variables(vars0);
+            wellSolutionState = state0;
+            variableStateExtractWellsVars(indices,vars,wellSolutionState);
+            asImpl().addWellEq(wellSolutionState, well_state, mob_perfcells_const, b_perfcells_const, aliveWells,cq_s);
+            addWellControlEq(wellSolutionState, well_state, aliveWells);
+            it++;
+            converged = getWellConvergence(it);
+        }
+        if (converged) {
+            std::cout << "well converged iter: " << it << std::endl;
+            const int nw = wells().number_of_wells;
+            {
+                // We will set the bhp primary variable to the new ones,
```

Comment:
I think perhaps these two blocks can be replaced by assigning
```
state.bhp = vars[0];
state.qs = vars[1];
```
inside the loop (or even using indices[Qs] and indices[Bhp] instead of 0 and 1). Could you test this?